### DISCUSSION TITLE
Prioritizing e2e test scenarios

### DISCUSSION THREAD

**Discussion by jdonenine** [k8ssandra/k8ssandra#703] [Ideas]
We've recently built up the capabilities within the project for e2e testing and we should now take advantage of that and continue building out scenarios that are of common/high priority.  To focus that effort I'd love to hear what some of those scenarios folks would really like to see covered are and get some sense of why they are important in one way or another.

To get the convo going here's a few ideas that have been sent my way or I'm recalling from various conversations recently:

1. Upgrade scenarios (e.g. existing 1.0.0 deployment upgraded via helm to 1.1.0), ideally in a generic fashion so that as the project progresses we can easily add tests for previous releases to the latest release.
2. Scale-up/down scenarios, where a cluster is deployed to a certain CassandraDatacenter size and it is then scaled up.  Perhaps this should include both projected success scenarios where sufficient resources are available as well as some, where the cluster cannot provision the additional resources necessary and what should be expected in both of those scenarios.

**Comment by jdonenine** [comment]
There were a few other ideas we had a shell of an idea put into the 1.0 testing plan we built up that I'll drop here as well, we'd need to flesh these ideas out a bit to prioritize them if anyone has any specific ideas on these:

* Rolling Restarts
* Cassandra Configuration Changes
* Negative Testing: Missing Storage Class
* Negative Testing: Size = 0

**Comment by jsanda** [comment]
I definitely think we want e2e test coverage for upgrade and scale up/down scenarios. Rolling restarts might be good as well.

I am not sure about C* config changes. We have unit test coverage for what is supported through k8ssandra. I think that may be sufficient. I don't think we would need to verify anything other than making sure changes are applied to the CassandraDatacenter. That is already done in the unit tests.

What would be the goal with testing a missing storage class? 

I am on the fence for testing size = 0. I tend to think that should be a scenario handled by cass-operator. The install will fail though due to an OpenAPI validation check. See [here](https://github.com/k8ssandra/cass-operator/blob/v1.6.0/operator/pkg/apis/cassandra/v1beta1/cassandradatacenter_types.go#L68).

**Reply by jdonenine** [reply]
> What would be the goal with testing a missing storage class?
>
>I am on the fence for testing size = 0. I tend to think that should be a scenario handled by cass-operator. The install will fail though due to an OpenAPI validation check. See here.

I'm not super sure on the goal/value here, I didn't add them personally, so something to discuss for sure.

For the size = 0 scenario, I'd guess we have some unit coverage for that scenario if we have the validation check on that already?
